File: frontend/src/api/overtime.ts
import api from './client'

export interface OvertimeItem {
  id: number
  user_id: number
  workflow_instance_id: number | null
  start_time: string
  end_time: string
  duration_hours: number
  reason: string
  status: string
  created_at: string
  updated_at: string
}

export interface PaginatedOvertimes {
  items: OvertimeItem[]
  total: number
  page: number
  page_size: number
}

export function overtimeStatusColor(status: string): string {
  const map: Record<string, string> = {
    draft: 'text-gray-500',
    submitted: 'text-blue-600',
    approved: 'text-green-600',
    rejected: 'text-red-600',
    cancelled: 'text-gray-400',
  }
  return map[status] || ''
}

export const overtimeApi = {
  list: (params?: { status?: string; page?: number; page_size?: number }) =>
    api.get<PaginatedOvertimes>('/overtimes', { params }),

  create: (data: {
    start_time: string
    end_time: string
    duration_hours: number
    reason: string
  }) => api.post<OvertimeItem>('/overtimes', data),

  getById: (id: number) => api.get<OvertimeItem>(`/overtimes/${id}`),

  update: (id: number, data: {
    start_time?: string
    end_time?: string
    duration_hours?: number
    reason?: string
  }) => api.put<OvertimeItem>(`/overtimes/${id}`, data),

  delete: (id: number) => api.delete(`/overtimes/${id}`),

  submit: (id: number) => api.post<OvertimeItem>(`/overtimes/${id}/submit`),

  cancel: (id: number) => api.post<OvertimeItem>(`/overtimes/${id}/cancel`),
}

File: backend/app/schemas/overtime.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class OvertimeCreate(BaseModel):
    start_time: datetime
    end_time: datetime
    duration_hours: float = Field(..., gt=0, le=168)
    reason: str = Field(..., min_length=1, max_length=2000)


class OvertimeUpdate(BaseModel):
    start_time: datetime | None = None
    end_time: datetime | None = None
    duration_hours: float | None = Field(None, gt=0, le=168)
    reason: str | None = Field(None, max_length=2000)


class OvertimeActualUpdate(BaseModel):
    actual_start_time: datetime | None = None
    actual_end_time: datetime | None = None


class OvertimeOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    user_id: int
    workflow_instance_id: int | None = None
    start_time: datetime
    end_time: datetime
    duration_hours: float
    actual_start_time: datetime | None = None
    actual_end_time: datetime | None = None
    reason: str
    status: str
    created_at: datetime
    updated_at: datetime


class PaginatedOvertimes(BaseModel):
    items: list[OvertimeOut]
    total: int
    page: int
    page_size: int
